### ISSUE TITLE
Leading comments incorrectly duplicated on variable initialization

### ISSUE THREAD

**Issue by kaicataldo** [eslint/espree#264]
Leading comments are incorrectly duplicated on variable initializations.  The current strategy employed to fix https://github.com/eslint/espree/issues/256 only checks the array inside `body` properties and removes comments that exist in the range of the previous index's node, never checking the `init` property (and so not covering this case). Just to be clear, this is not a regression - this is just one more case that was being added incorrectly (or rather, in practice, not being removed correctly). 

The difficulty with this one is that when we drill deeper into the tree, we'll need to have saved some kind of reference to the previously iterated over comments to know if we need to remove it or not. We may have to rethink our strategy on this.

Example:
The following code
```js
//foo
var zzz /*aaa*/ = 777;
//bar
```
currently produces this output (notice the incorrect leading comment `//foo` on the literal `777`):
```js
{
    "type": "Program",
    "start": 0,
    "end": 34,
    "range": [
        6,
        28
    ],
    "body": [
        {
            "type": "VariableDeclaration",
            "start": 6,
            "end": 28,
            "range": [
                6,
                28
            ],
            "declarations": [
                {
                    "type": "VariableDeclarator",
                    "start": 10,
                    "end": 27,
                    "range": [
                        10,
                        27
                    ],
                    "id": {
                        "type": "Identifier",
                        "start": 10,
                        "end": 13,
                        "range": [
                            10,
                            13
                        ],
                        "name": "zzz",
                        "trailingComments": [
                            {
                                "type": "Block",
                                "value": "aaa",
                                "start": 14,
                                "end": 21,
                                "range": [
                                    14,
                                    21
                                ]
                            }
                        ]
                    },
                    "init": {
                        "type": "Literal",
                        "start": 24,
                        "end": 27,
                        "range": [
                            24,
                            27
                        ],
                        "value": 777,
                        "raw": "777",
                        "leadingComments": [
                            {
                                "type": "Line",
                                "value": "foo",
                                "start": 0,
                                "end": 5,
                                "range": [
                                    0,
                                    5
                                ]
                            },
                            {
                                "type": "Block",
                                "value": "aaa",
                                "start": 14,
                                "end": 21,
                                "range": [
                                    14,
                                    21
                                ]
                            }
                        ]
                    }
                }
            ],
            "kind": "var",
            "leadingComments": [
                {
                    "type": "Line",
                    "value": "foo",
                    "start": 0,
                    "end": 5,
                    "range": [
                        0,
                        5
                    ]
                }
            ],
            "trailingComments": [
                {
                    "type": "Line",
                    "value": "bar",
                    "start": 29,
                    "end": 34,
                    "range": [
                        29,
                        34
                    ]
                }
            ]
        }
    ],
    "sourceType": "script",
    "comments": [
        {
            "type": "Line",
            "value": "foo",
            "start": 0,
            "end": 5,
            "range": [
                0,
                5
            ]
        },
        {
            "type": "Block",
            "value": "aaa",
            "start": 14,
            "end": 21,
            "range": [
                14,
                21
            ]
        },
        {
            "type": "Line",
            "value": "bar",
            "start": 29,
            "end": 34,
            "range": [
                29,
                34
            ]
        }
    ]
}
```

I plan to look at this.

**Comment by kaicataldo** [comment]
Haven't forgotten about this. It's been on my mind and I hope to get to it soon.

**Comment by kaicataldo** [comment]
After trying a few iterations (and trying to write the comment attachment from scratch), I have a much better understanding of why the current system for comment attachment works the way it does. With that newfound understanding, I think I have a much more elegant solution than my original (recursively removing comments that occur before the previous node's start location at the Program node level) that also fixes this issue. 

I know we talked about moving comment attachment to ESLint, and I'm definitely willing to give that a shot - wanted to fix this issue here first at the parser level (we can always just not use the `attachComment` option if we decide to try to attach comments at the ESLint level).

The branch with my changes has a few failing tests when parsing the libraries - most likely from having duplicate comments attached to nested nodes - but once I make sure everything is working and can fix the tests I'll open a PR.

**Comment by kaicataldo** [comment]
Sorry, one last bit of info. Changing the test's expected order of events from
 ```js
var expected = [
    ["Program", ast],
    ["LineComment", ast.comments[0]], // foo
    ["VariableDeclaration", ast.body[0]],
    ["LineComment", ast.comments[2]], // bar
    ["VariableDeclarator", ast.body[0].declarations[0]],
    ["Identifier", ast.body[0].declarations[0].id],
    ["BlockComment", ast.comments[1]], /* aaa */
    ["BlockComment:exit", ast.comments[1]], /* aaa */
    ["Identifier:exit", ast.body[0].declarations[0].id],
    ["Literal", ast.body[0].declarations[0].init],
    ["Literal:exit", ast.body[0].declarations[0].init],
/ * Move this --> */ ["LineComment:exit", ast.comments[0]], // foo
    ["VariableDeclarator:exit", ast.body[0].declarations[0]],
    ["LineComment:exit", ast.comments[2]], // bar
    ["VariableDeclaration:exit", ast.body[0]],
    ["Program:exit", ast]
];
```
to
```js
var expected = [
    ["Program", ast],
    ["LineComment", ast.comments[0]], // foo
    ["VariableDeclaration", ast.body[0]],
    ["LineComment", ast.comments[2]], // bar
    ["VariableDeclarator", ast.body[0].declarations[0]],
    ["Identifier", ast.body[0].declarations[0].id],
    ["BlockComment", ast.comments[1]], /* aaa */
    ["BlockComment:exit", ast.comments[1]], /* aaa */
    ["Identifier:exit", ast.body[0].declarations[0].id],
    ["Literal", ast.body[0].declarations[0].init],
    ["Literal:exit", ast.body[0].declarations[0].init],
    ["VariableDeclarator:exit", ast.body[0].declarations[0]],
    ["LineComment:exit", ast.comments[2]], // bar
    ["VariableDeclaration:exit", ast.body[0]],
/* To here --> */ ["LineComment:exit", ast.comments[0]], // foo
    ["Program:exit", ast]
];
```
causes the failing test to pass. Sorry for the massive amount of info, and really appreciate any guidance on the issue!

**Comment by kaicataldo** [comment]
@nzakas Pinging once more - let me know if you'd prefer we don't ping multiple times (don't want to bug you, but also wanted to make sure this didn't get lost). I know you've been super busy with logistical stuff this week (but yay for all the exciting things!), so whenever you get a chance :)
